nucliadb_vectors/src/data_point/ops_hnsw.rs

```diff
@@ -129,37 +128,54 @@ impl<'a, DR: DataRetriever> HnswOps<'a, DR> {
         let picked_level = -sample.ln() * params::level_factor();
         picked_level.round() as usize
     }
-    fn closest_up_node<L: Layer>(
+    fn closest_up_nodes<L: Layer>(
         &'a self,
-        x: Address,
+        entry_points: Vec<Address>,
         query: Address,
         layer: L,
-        filter: NodeFilter<'a, DR>,
-    ) -> Option<(Address, f32)> {
+        number_of_results: usize,
+        mut filter: NodeFilter<'a, DR>,
+    ) -> Vec<(Address, f32)> {
         // We just need to perform BFS, the replacement is the closest node to the actual
         // best solution. This algorithm takes a lazy approach to computing the similarity of
         // candidates.
-        let mut visited_nodes = HashSet::new();
-        let mut candidates = VecDeque::from([x]);
+
+        let mut results = Vec::new();
+        let mut visited_nodes: HashSet<_> = HashSet::from_iter(entry_points.iter().copied());
```

Comment:
You might want to use a bit-set here. I tried an approach very similar to this as well and it had the same problem where the `HashSet` represented 30% of the runtime. So if this is now taking 2s, switching to a bitfield might lower it to 1.5s

Reply:
You are right! This took down our benchmark results to 1.15 secs!! 🤯